File: codility/lessons/lesson03_Time-Complexity/TapeEquilibrium.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int solution(int A[], int N) {
	int i;
	long sum;
	long acc;
	long min = INT_MAX;

	for (sum = 0, i = 0; i < N; i++) {
		sum += (long)(A[i]);
	}

	for (acc = 0, i = 0; i < N - 1; i++) {
		long diff;

		acc += (long)(A[i]);
		sum -= (long)(A[i]);
		//printf("%ld - %ld\n", sum, acc);
		diff = sum - acc;
		if (diff < 0) {
			diff *= -1l;
		}

		if (diff < min) {
			min = diff;
		}
	}

	return (int)min;
}

int main(void) {
	int A[] = {3, 1, 2, 4, 3};

	printf("%d\n", solution(A, 5));

	return EXIT_SUCCESS;
}

File: codility/lessons/lesson03_Time-Complexity/TapeEquilibrium.py
# https://app.codility.com/programmers/lessons/3-time_complexity/tape_equilibrium/

def solution(A):
    lsum = A[0]
    rsum = sum(A[1:])
    ans = abs(lsum - rsum)
    for a in A[1:-1]:
        lsum += a
        rsum -= a
        ans = min(ans, abs(lsum - rsum))
        print(ans)

    return ans


cases = [
        #([3,1,2,4,3], 1),
        ([-1000,1000], 2000),
        ]
for a, gt in cases:
    assert solution(a) == gt
